pkg/controller/sidecarset/sidecarset_hotupgrade.go

```diff
@@ -0,0 +1,227 @@
+	}
+	return false
+}
+
+func updateHotUpgradeContainerInPod(sidecarContainer *appsv1alpha1.SidecarContainer, control sidecarcontrol.SidecarControl, pod *corev1.Pod) (changedContainer string) {
+	sidecarSet := control.GetSidecarset()
+	containerInPods := make(map[string]corev1.Container)
+	for _, containerInPod := range pod.Spec.Containers {
+		containerInPods[containerInPod.Name] = containerInPod
+	}
+
+	// the container will be upgraded in latest sidecarSet specification
+	nameToUpgrade := findContainerToHotUpgrade(sidecarContainer, pod, control)
+	containerToUpgrade := containerInPods[nameToUpgrade]
+	newContainer := control.UpdateSidecarContainerToLatest(sidecarContainer.Container, containerToUpgrade)
+	// new hot upgrade sidecar container specification
+	afterContainerSpec := util.DumpJSON(newContainer)
+
+	// older hot upgrade sidecar container specification
+	var beforeContainerSpec string
+	name1, name2 := sidecarcontrol.GetHotUpgradeContainerName(sidecarContainer.Name)
+	if nameToUpgrade == name1 {
+		beforeContainerSpec = util.DumpJSON(containerInPods[name2])
+	} else {
+		beforeContainerSpec = util.DumpJSON(containerInPods[name1])
+	}
+
+	// sidecarToUpgrade: sidecarSet.Spec.Container[x].name -> sidecar container in pod
+	// for example: mesh -> mesh-1, envoy -> envoy-2...
+	sidecarToUpgrade := make(map[string]string)
+	// pod.container definition changed, then update container spec in pod
+	if beforeContainerSpec != afterContainerSpec {
+		klog.V(3).Infof("try to update container %v/%v/%v, before: %v, after: %v",
+			pod.Namespace, pod.Name, newContainer.Name, beforeContainerSpec, afterContainerSpec)
+		updateContainerInPod(newContainer, pod)
+		sidecarToUpgrade[sidecarContainer.Name] = newContainer.Name
+		sidecarcontrol.RecordHotUpgradeInfoInAnnotations(sidecarToUpgrade, pod)
+		// update sidecar container resource version in annotations
+		pod.Annotations[sidecarcontrol.GetSidecarContainerVersionAnnotationKey(newContainer.Name)] = sidecarSet.ResourceVersion
+		changedContainer = newContainer.Name
+	}
+	return
+}
+
+func findContainerToHotUpgrade(sidecarContainer *appsv1alpha1.SidecarContainer, pod *corev1.Pod, control sidecarcontrol.SidecarControl) string {
+	containerInPods := make(map[string]corev1.Container)
+	for _, containerInPod := range pod.Spec.Containers {
+		containerInPods[containerInPod.Name] = containerInPod
+	}
+	name1, name2 := sidecarcontrol.GetHotUpgradeContainerName(sidecarContainer.Name)
+	c1, c2 := containerInPods[name1], containerInPods[name2]
+
+	// First, empty hot sidecar container will be upgraded with the latest sidecarSet specification
+	if c1.Image == sidecarContainer.UpgradeStrategy.HotUpgradeEmptyImage {
+		return c1.Name
+	} else if c2.Image == sidecarContainer.UpgradeStrategy.HotUpgradeEmptyImage {
+		return c2.Name
+	}
+
+	// Second, Not ready sidecar container will be upgraded
+	c1Ready, c2Ready := isContainerConsistentAndReady(&c1, pod, control), isContainerConsistentAndReady(&c2, pod, control)
```

Comment:
Can I assume that if the code reaches here, it means the previous hot upgrade has not finished yet?  If we continue to update, is it possible to cause "working" sidecar container to be recreated, hence breaking the service?

Reply:
In this scenario, if the last hot upgrade fails (the new version of sidecar does start to fail) and the user adjusts the image and then updates the SidecarSet to trigger the upgrade, the container that is not ready will be selected first to upgrade

Reply:
For the scenario you said above I made a judgment in the **IsSidecarSetCanUpgrade** method in the file [sidecarset_control.go](https://github.com/zmberg/kruise/blob/sidecarset_hotupgrade/pkg/control/sidecarcontrol/sidecarset_control.go), for being in the process of upgrading pod this time I will not upgrade.